### PR TITLE
Prototype for DQDM.

### PR DESCRIPTION
**Pull request by trivialfis**
* Add `DeviceQuantileDMatrix`.  This class accepts only iterator and does not allow individual meta data setter.
* Remove `DataFrameBatch` as it's unnecessary.
* Accept actual data frames instead of column indices in `ColumnBatch`.
* Move `getArrayInterfaceJson` into the base class and let the subclasses define getters for each data field like `labels` and `weights`.  This way we can modify the internal of this function whenever needed.

### CODE REVIEW

**Comment by trivialfis** [comment]
I have converted all the related tests.  Will mark to-do items in inlined comments.

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/CudfColumnBatch.java [hunk 1]
```diff
@@ -28,148 +28,68 @@
...
-
-  /**
-   * CudfColumnBatch constructor
-   * @param table             the CUDF table
-   * @param featureIndices    must-have, specify the feature's indices in the table
-   */
-  public CudfColumnBatch(Table table, int[] featureIndices) {
-    this(table, featureIndices, null, null, null);
+  private final Table labels;
+  private final Table weights;
```
Comment (trivialfis) [inline jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/CudfColumnBatch.java:31]: TODO: Maybe use `ColumnVector` or `Column` instead?

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/CudfColumnBatch.java [hunk 2]
```diff
@@ -28,148 +28,68 @@
...
-  private CudfColumn[] getAsCudfColumn(int... indices) {
-    if (indices == null || indices.length == 0) return new CudfColumn[]{};
+  private CudfColumn[] getAsCudfColumn(Table data) {
+    if (data == null) {
+      return new CudfColumn[]{};
+    }
+    int[] indices = this.buildIndices(data);
     return Arrays.stream(indices)
-      .mapToObj(this::getColumnVector)
+      .mapToObj((i) -> data.getColumn(i))
```
Comment (trivialfis) [inline jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/CudfColumnBatch.java:86]: TODO: I hope lambda is fine.

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/DeviceQuantileDMatrix.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        iter, missing, maxBin, nthread, out));
+    handle = out[0];
+  }
+
+  @Override
+  public void setLabel(Column column) throws XGBoostError {
+    throw new XGBoostError("Not applicable.");
+  }
+  @Override
+  public void setWeight(Column column) throws XGBoostError {
```
Comment (trivialfis) [inline jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/DeviceQuantileDMatrix.java:29]: TODO: Other fields.

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/ColumnBatch.java [hunk 4]
```diff
@@ -41,7 +64,12 @@
    */
   public abstract String getFeatureArrayInterface();
 
+  public abstract String getLabelsArrayInterface();
```
Comment (trivialfis) [inline jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/ColumnBatch.java:67]: TODO: Documents.

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/ColumnBatch.java [hunk 5]
```diff
@@ -16,23 +16,46 @@
...
    * Get the cuda array interface json string for the whole ColumnBatch including
    * the must-have feature, label columns and the optional weight, base margin columns.
    *
-   * This API will be called by {@link DMatrix#DMatrix(Iterator, float, int, int)}
-   *
+   * This function is be called by native code during iteration and can be made as private
+   * method.  We keep it as public simply to silent the linter.
    */
-  public abstract String getArrayInterfaceJson();
+  public final String getArrayInterfaceJson() {
```
Comment (trivialfis) [inline jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/ColumnBatch.java:31]: The function is considered to be internal and can be replaced by a C++ implementation.  The important part is the JSON format we pass into libxgboost.so can be changed whenever needed.

File: jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/BoosterTest.java [hunk 6]
```diff
@@ -83,7 +85,16 @@ public void testBooster() throws XGBoostError {
 
     try (Table tmpTable = Table.readCSV(schema, opts,
 					new File(trainingDataPath))) {
-      CudfColumnBatch batch = new CudfColumnBatch(tmpTable, new int[]{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}, new int[]{12});
+      ColumnVector[] df = new ColumnVector[12];
+      for (int i = 0; i < 12; ++i) {
```
Comment (RAMitchell) [inline jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/BoosterTest.java:89]: Is there a functional way to do this?

File: jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/BoosterTest.java [hunk 7]
```diff
@@ -83,7 +85,16 @@ public void testBooster() throws XGBoostError {
...
+        df[i] = tmpTable.getColumn(i);
+      }
+      Table X = new Table(df);
+      ColumnVector[] labels = new ColumnVector[1];
+      labels[0] = tmpTable.getColumn(12);
+      Table y = new Table(labels);
+
+      CudfColumnBatch batch = new CudfColumnBatch(X, y, null, null);
```
Comment (RAMitchell) [inline jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/BoosterTest.java:97]: What's the necessity of the `CudfColumnBatch` class? Could `DeviceQuantileDMatrix` directly accept this object `ColumnVector[] df`?

Comment (trivialfis) [reply jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/BoosterTest.java:97]: DQDM accepts only iterators.  The `CudfColumnBatch` is an adapter.  I have made a similar suggestion to @wbo4958 so I'm open to consuming `Iterator<Table>` directly and move DQDM into xgboost4j-gpu package (so we don't bring cuDF into CPU package's dependency tree).

File: jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/BoosterTest.java (hunk 6 above)
Comment (trivialfis) [reply jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/BoosterTest.java:89]: I glanced through the java binding of cuDF, there's a `getColumns` method but it's private.  @wbo4958 might know better than I do.

File: jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/BoosterTest.java (hunk 7 above)
Comment (RAMitchell) [reply jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/BoosterTest.java:97]: Is it possible for DQDM to accept different iterators for labels and feature columns? e.g. `DeviceQuantileDMatrix(features.iterator(), labels.iterator(), Float.NaN, maxBin, 1);`

Comment (trivialfis) [reply jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/BoosterTest.java:97]: But that also complicates the construction of normal DMatrix, which cannot have cudf table in its definition due to dependency issues.

Comment (trivialfis) [reply jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/BoosterTest.java:97]: @RAMitchell 

> Is it possible for DQDM to accept different iterators for labels and feature columns?

Yes, but with some other caveats.  The first possible approach is we pass down all the iterators into libxgboost.so, but we will have a weird and inflexible C API that accepts many iterator pointers.  The second approach would be we concatenate all the iterators into one in java, but that's effectively the same as current implementation except that we do it for the users.

**Pull request merged**